pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,84 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	logger "github.com/rs/zerolog/log"
+)
+
+var (
+	c           *collector
+	metricStore MetricStore
+	log         zerolog.Logger
+
+	mu   = &sync.Mutex{}
+	once sync.Once
+)
+
+func init() {
+	log = logger.With().
+		Str("component", "telemetry").
+		Logger()
+}
+
+// MetricStore specifies the methods for persisting a metric.
+type MetricStore interface {
+	StoreMetric(context.Context, Metric) error
+	Close() error
+}
+
+// SetMetricStore sets the store implementation.
+// Can only be called once, and should be called before Collect.
+func SetMetricStore(s MetricStore) {
+	once.Do(func() {
+		metricStore = s
+	})
+}
+
+// Collect collects the metric by persisting locally for later publication.
+// If Collect is called before setting the metric store, it will simply log the metric without persisting it.
+func Collect(ctx context.Context, metric interface{}) error {
+	mu.Lock()
+	defer mu.Unlock()
+	if metricStore == nil {
+		log.Warn().Msg("no metric store was set")
+		return nil
+	}
+
+	if c == nil {
+		c = &collector{
+			s: metricStore,
+		}
+	}
+
+	switch v := metric.(type) {
```

Comment:
This is another decision I made. A metric to satisfy a specific set of interfaces. In this case, we only have only: `StateHash`.  I could have let the door open, and any structure could be considered a metric. However, I thought letting the door open could lead to a case where we have too much garbage in the metrics database. With this restriction, we can be more thoughtful about which kind of data we add to the metrics database.